nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/DecryptMode.groovy

```diff
@@ -0,0 +1,322 @@
+        cli.k(longOpt: 'key',
+                args: 1,
+                argName: 'keyhex',
+                optionalArg: true,
+                'Use a raw hexadecimal key to decrypt the input file. If an argument is not provided to this flag, interactive mode will be triggered to prompt the user to enter the key.')
+        cli.b(longOpt: 'bootstrapConf',
+                args: 1,
+                argName: 'file',
+                'Use a bootstrap.conf file containing the master key to decrypt the input file (as an alternative to -p or -k)')
+
+        cli.o(longOpt: 'output',
+                args: 1,
+                argName: 'file',
+                'Specify an output file. If omitted, Standard Out is used. Output file can be set to the input file to decrypt the file in-place.')
+
+        return cli
+
+    }
+
+    static class DecryptConfiguration {
+
+        OptionAccessor rawOptions
+
+        Configuration.KeySource keySource
+        String key
+        SensitivePropertyProvider decryptionProvider
+        String inputBootstrapPath
+
+        FileType fileType
+
+        String inputFilePath
+
+        boolean outputToFile = false
+        String outputFilePath
+
+        DecryptConfiguration() {
+        }
+
+        DecryptConfiguration(OptionAccessor options) {
+            this.rawOptions = options
+
+            validateOptions()
+            determineInputFileFromRemainingArgs()
+
+            determineKey()
+            if (!key) {
+                throw new RuntimeException("Failed to configure tool, could not determine key.")
+            }
+            decryptionProvider = new AESSensitivePropertyProvider(key)
+
+            if (rawOptions.t) {
+                fileType = FileType.valueOf(rawOptions.t)
+            }
+
+            if (rawOptions.o) {
+                outputToFile = true
+                outputFilePath = rawOptions.o
+            }
+        }
+
+        private void validateOptions() {
+
+            String validationFailedMessage = null
+
+            if (!rawOptions.b && !rawOptions.p && !rawOptions.k) {
+                validationFailedMessage = "-p, -k, or -b is required in order to determine the master key to use for decryption."
+            }
+
+            if (validationFailedMessage) {
+                throw new RuntimeException("Invalid options: " + validationFailedMessage)
+            }
+
+        }
+
+        private void determineInputFileFromRemainingArgs() {
+            String[] remainingArgs = this.rawOptions.getInner().getArgs()
+            if (remainingArgs.length == 0) {
+                throw new RuntimeException("Missing argument: Input file must provided.")
+            } else if (remainingArgs.length > 1) {
+                throw new RuntimeException("Too many arguments: Please specify exactly one input file in addition to the options.")
```

Comment:
Is order important here? If `$ encrypt-config.sh --decrypt -b bootstrap.conf file_to_decrypt.properties` works but `$ encrypt-config.sh file_to_decrypt.properties --decrypt -b bootstrap.conf` doesn't, I think this message should indicate that.

Reply:
This is actually correct if you call `DecryptMode.run(String[])` (not exposed to end users in this commit). The usage you're thinking of is through `NiFiRegistryDecryptMode.run(String[])`, which does its own argument parsing and calls `DecryptMode.run(DecryptConfiguration)`